Source/relay/TourGuide/Items of the Month/2023/Closed Circuit Pay Phone.ash

```diff
@@ -0,0 +1,123 @@
+    // (but not yet called Rufus back)
+    state.state_boolean["quest objective fulfilled"] = state.mafia_internal_step == 2;
+
+    return state;
+}
+
+RegisterTaskGenerationFunction("IOTMClosedCircuitPayPhoneGenerateTasks");
+void IOTMClosedCircuitPayPhoneGenerateTasks(ChecklistEntry [int] task_entries, ChecklistEntry [int] optional_task_entries, ChecklistEntry [int] future_task_entries) {
+    if (!lookupItem("closed-circuit pay phone").have())
+        return;
+
+    string url = "inv_use.php?pwd=" + my_hash() + "&which=3&whichitem=11169";
+    QuestState state = parseRufusQuestState();
+
+    ChecklistEntry [int] whereToAddRufusQuestTile;
+    string rufusImage = "__item closed-circuit pay phone";
+    string rufusQuestTitle;
+    string [int] rufusQuestDescription;
+    int rufusQuestPriority;
+    
+    int shadowLodestones = available_amount($item[Rufus's shadow lodestone]);
+    if (shadowLodestones > 0) {
+        rufusQuestDescription.listAppend(HTMLGenerateSpanFont("Have " + pluralise($item[Rufus's shadow lodestone]) + ".", "purple"));
+    }
+
+    int riftAdvsUntilNC = get_property_int("encountersUntilSRChoice");
+    rufusQuestDescription.listAppend(HTMLGenerateSpanFont(riftAdvsUntilNC + " encounters until NC/boss.", "black"));
+
+    if (state.state_boolean["quest objective fulfilled"]) {
+        // We've fulfilled the quest objective but still need to call Rufus
+        rufusQuestDescription.listAppend(HTMLGenerateSpanFont("Call Rufus and get a lodestone", "black"));
+        rufusQuestTitle = "Rufus quest done";
+        rufusQuestPriority = -11;
+        whereToAddRufusQuestTile = task_entries;
+    }
+    else if (state.started && riftAdvsUntilNC == 0) {
+        rufusQuestDescription.listAppend(HTMLGenerateSpanFont("Fight a boss or get an artifact", "black"));
+        rufusQuestTitle = "Shadow Rift NC up next";
+        rufusQuestPriority = -11;
+        rufusImage = "__item shadow bucket";
+        whereToAddRufusQuestTile = task_entries;
+    }
+    else if (state.started) {
+        rufusQuestTitle = "Rufus quest in progress";
+        rufusQuestPriority = 11;
+        whereToAddRufusQuestTile = optional_task_entries;
+    }
+    else if (!state.started) {
+        boolean calledRufusToday = get_property_boolean("_shadowAffinityToday");
+        string textColor = calledRufusToday ? "black" : "red";
+        string callRufusMessage = calledRufusToday ? "Optionally call Rufus again for another (turn-taking) quest." : "Haven't called Rufus yet today.";
+        rufusQuestDescription.listAppend(HTMLGenerateSpanFont(callRufusMessage, textColor));
+        rufusQuestTitle = "Rufus quest doable now";
+        rufusQuestPriority = 11;
+        whereToAddRufusQuestTile = optional_task_entries;
+    }
+
+    whereToAddRufusQuestTile.listAppend(ChecklistEntryMake(rufusImage, url, ChecklistSubentryMake(rufusQuestTitle, "", rufusQuestDescription), rufusQuestPriority));
```

Comment:
i like this logic change. much easier to read and easier to change later on. thanks zinc.